### DISCUSSION TITLE
Parsing logic

### DISCUSSION THREAD

**Discussion by tetotechy** [toml-lang/toml#992] [Q&A]
Hi all,

far from being an abnf expert, but trying to implement a parser for toml files I got stuck in what apparently was a bug in my code and now I wonder whether it is something else. The line under investigation is [this](https://github.com/toml-lang/toml/blob/8eae5e1c005bc5836098505f85a7aa06568999dd/toml.abnf#L20). Both terms here (`ws` and `[ comment ]`) are optional so my parser got stuck forever on this line for a file that started with neither whitespace nor a comment at the beginning. This, because the line is satisfied by `""`, and the parser never steps ahead to the first character.

**Comment by ChristianSi** [comment]
Technically that's not wrong. The shortest possible TOML document is supposedly just an empty string which is matched by this rule, as it should be.

However, you should change your parser to by "greedy", that is, trying to match as much as possible instead of as little as possible.
